dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/OrganisationUnitObjectBundleHook.java

```diff
@@ -101,13 +113,63 @@ public void preUpdate( OrganisationUnit object, OrganisationUnit persistedObject
     public void validate( OrganisationUnit organisationUnit, ObjectBundle bundle,
         Consumer<ErrorReport> addReports )
     {
-        if ( organisationUnit.getClosedDate() != null
-            && organisationUnit.getClosedDate().before( organisationUnit.getOpeningDate() ) )
+        validateOpeningClosingDate( organisationUnit, addReports );
+        validatePotentialMove( organisationUnit, bundle, addReports );
+    }
+
+    private void validateOpeningClosingDate( OrganisationUnit object, Consumer<ErrorReport> addReports )
+    {
+        if ( object.getClosedDate() != null && object.getClosedDate().before( object.getOpeningDate() ) )
+        {
+            addReports.accept( new ErrorReport( OrganisationUnit.class, ErrorCode.E4013,
+                object.getClosedDate(), object.getOpeningDate() ) );
+        }
+    }
+
+    private void validatePotentialMove( OrganisationUnit unit, ObjectBundle bundle, Consumer<ErrorReport> addReports )
+    {
+        User user = bundle.getUser();
+        if ( user == null || user.isSuper() || !bundle.isPersisted( unit ) )
+        {
+            return; // not an update or always permitted for superuser
+        }
+        OrganisationUnit oldParent = bundle.getPreheat().get( bundle.getPreheatIdentifier(), unit ).getParent();
+        OrganisationUnit newParent = unit.getParent();
+        if ( Objects.equals( getNullableUid( oldParent ), getNullableUid( newParent ) ) )
         {
-            addReports
-                .accept( new ErrorReport( OrganisationUnit.class, ErrorCode.E4013, organisationUnit.getClosedDate(),
-                    organisationUnit.getOpeningDate() ) );
+            return; // not a move
         }
+        if ( !user.isAuthorized( "F_ORGANISATIONUNIT_MOVE" ) )
+        {
+            addReports.accept(
+                new ErrorReport( OrganisationUnit.class, ErrorCode.E1515, user.getUid() ) );
+            return;
+        }
+        if ( !aclService.canWrite( user, unit ) )
+        {
+            addReports.accept( new ErrorReport( OrganisationUnit.class, ErrorCode.E1516, user.getUid(),
+                getUidOrName( unit ) ) );
+        }
+        if ( oldParent != null && !aclService.canWrite( user, oldParent ) )
+        {
+            addReports.accept( new ErrorReport( OrganisationUnit.class, ErrorCode.E1517, user.getUid(),
```

Comment:
If the user doesn't have write access to the old parent, it could be confusing to say that the user cannot move "to" the old parent. One way to prevent this confusion would be to add another error message:

    E1515( "User `{0}` is not allowed to move organisation units" ),
    E1516( "User `{0}` is not allowed to move organisation `{1}`" ),
    E1517( "User `{0}` is not allowed to move organisation `{1}` unit from parent `{2}`" ),
    E1518( "User `{0}` is not allowed to move organisation `{1}` unit to parent `{2}`" ),

Otherwise, looks great!

Reply:
Good spot. This wasn't intentional. Just didn't think this through properly :D. I'll make the change as you suggested on Monday.

Reply:
Done.